api/tools/src/main/java/org/pmiops/workbench/tools/ApplicationRunListener.java

```diff
@@ -0,0 +1,71 @@
+package org.pmiops.workbench.tools;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.SpringApplicationRunListener;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.context.support.SimpleThreadScope;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.web.context.WebApplicationContext;
+
+/**
+ * A custom application listener to allow us to mutate the application context before any
+ * command-line tools are run.
+ *
+ * <p>This class currently serves a single purpose, which is to create a fake "request" bean scope
+ * to allow command-line tools to load RW beans that are request-scoped for use in the main webapp.
+ *
+ * <p>See https://stackoverflow.com/a/28275111 which is the pattern we roughly followed here. This
+ * class is referenced from resources/META-INF/spring.factories, which causes this class to be
+ * loaded before the Spring context is initialized.
+ */
+public class ApplicationRunListener implements SpringApplicationRunListener {
+
+  public ApplicationRunListener(SpringApplication application, String[] args) {}
+
+  /**
+   * Called immediately when the run method has first started. Can be used for very early
+   * initialization.
+   */
+  @Override
+  public void starting() {}
+
+  /**
+   * Called once the environment has been prepared, but before the {@link ApplicationContext} has
+   * been created.
+   *
+   * @param environment the environment
+   */
+  @Override
+  public void environmentPrepared(ConfigurableEnvironment environment) {}
+
+  /**
+   * Called once the {@link ApplicationContext} has been created and prepared, but before sources
+   * have been loaded.
+   *
+   * @param context the application context
+   */
+  @Override
+  public void contextPrepared(ConfigurableApplicationContext context) {
+    context
+        .getBeanFactory()
+        .registerScope(WebApplicationContext.SCOPE_REQUEST, new SimpleThreadScope());
```

Comment:
Also, I'd consider extending `SimpleThreadScope` with something like `AoUPreConfigRequestScopePlaceholder` or otherwise suspicious and using that so it shows up in stack traces or config graphs, which would increase the likelihood of a maintainer finding this.

Reply:
Could be nice, but sounds pretty hypothetical. I'd rather not end up with more files floating around here.